File: backend/api/dns_bypass.py
"""FastAPI-Router des netzweiten DNS-Umgehungs-Waechters (ADR 0042, Etappe 4b), prefix
``/api/dns-bypass``.

Aeusserer Ring: nimmt HTTP entgegen. Wie ``api/dns_watch.py`` (und ``api/outbound.py``)
kennt dieser Router WEDER ``application`` NOCH ``infrastructure`` NOCH ``modules``
(Regel 4): die injizierten Runner kommen als schmale lokale Vertraege per Dependency
herein, verdrahtet im Composition Root (``app.py``).

Damit der api-Ring application-frei bleibt UND der Aufruf trotzdem typsicher ist (mypy
strict), beschreibt je ein schmales lokales ``Protocol``/``Callable`` den Vertrag der
injizierten Runner (Lesen der verdichteten Sicht, billiger Status-Poll, Start/Stop der
Aufzeichnung). Der api-Ring definiert eigene schmale pydantic-Response-Modelle
(``DnsBypassFindingOut``/``DnsBypassOverviewOut``/``DnsBypassStatusOut``); die Projektion
vom application-Typ ``DnsBypassOverview`` auf diese Wire-Form UND die Geraete-/DoH-
Zuordnung machen die Composition-Root-Runner in ``app.py``, NICHT der Router -- so nennt
der api-Ring den application-Typ nie.

Endpunkte:

* ``GET /api/dns-bypass`` -> die verdichtete netzweite Umgehungs-Sicht (je Umgehung ein
  Befund mit best-effort Geraetename + DoH-Bewertung, plus Zaehler und ``recording``).
* ``GET /api/dns-bypass/status`` -> billiger Status-Poll (laeuft die Aufzeichnung? wie
  viele Anfragen sind gesammelt?), ohne die teure Verdichtung.
* ``POST /api/dns-bypass/start`` -> startet die Aufzeichnung ON-DEMAND (Body
  ``{interface?}``). Der Helfer-Start kann legitim scheitern (z. B. keine Rechte): ein
  Fehlertext wird als ``{"ok": false, "error": <text>}`` mit HTTP 200 ehrlich
  durchgereicht (kein 500 -- S3, kein stiller Fallback).
* ``POST /api/dns-bypass/stop`` -> stoppt die Aufzeichnung. ``{"ok": true}``.
"""

from collections.abc import Callable
from typing import Annotated, Protocol

from fastapi import APIRouter, Depends
from pydantic import BaseModel

router = APIRouter(prefix="/api/dns-bypass", tags=["dns_bypass"])


# ── schmale api-Response-Modelle (eigene Wire-Form, KEIN application-Typ) ──────
# Die Felder spiegeln ``application.dns_bypass.DnsBypassFinding``/``DnsBypassOverview``,
# ohne diese Typen zu importieren. Der Composition-Root-Runner projiziert die
# application-Sicht auf genau diese Form und reichert sie um Geraetename + DoH-Bewertung
# an (Regel 4/5: api kennt application nicht; die Zuordnung faellt nur im Root).


class DnsBypassFindingOut(BaseModel):
    """Ein aggregierter Umgehungs-Befund = EINE (``src_ip``, ``dst_ip``)-Gruppe.

    ``device_name`` ist der best-effort im Bestand gefundene Anzeigename der Quell-IP
    (``None`` wenn keine IP passt -- Option 2, kein harter Identitaets-Anspruch);
    ``is_doh``/``doh_source_name`` sind die im Composition Root ueber die eigene
    DoH-Lookup-Naht ermittelte Bewertung des Ziels (Quellenname der treffenden aktiven
    DOH-Quelle, sonst ``None``).
    """

    src_ip: str
    device_name: str | None
    dst_ip: str
    is_doh: bool
    doh_source_name: str | None
    query_count: int
    sample_qnames: list[str]


class DnsBypassOverviewOut(BaseModel):
    """Die verdichtete netzweite Umgehungs-Sicht: alle Befunde + Zaehler + ``recording``.

    ``expected_servers`` ist die zur Klassifikation genutzte erwartete-Resolver-Menge
    (ehrlicher Beleg, was der Befund bedeutet). ``recording`` sagt ehrlich, ob die
    Aufzeichnung gerade laeuft (ein Leerbefund bei ``recording=false`` bedeutet: es wird
    nicht mitgelesen, nicht "alles sauber").
    """

    findings: list[DnsBypassFindingOut]
    expected_servers: list[str]
    queries_total: int
    bypass_total: int
    expected_total: int
    bypass_devices: int
    recording: bool


class DnsBypassStatusOut(BaseModel):
    """Billiger Status-Poll: laeuft die Aufzeichnung? wie viele Anfragen sind gesammelt?

    Bewusst OHNE die teure Verdichtung (``BuildDnsBypass``) -- nur der laufende Zustand
    des Recorders, damit die UI guenstig pollen kann.
    """

    recording: bool
    collected_queries: int


# ── injizierte Composition-Root-Runner ────────────────────────────────────────
# Provider-Marker: in app.py per dependency_overrides mit den echten Root-Runnern
# verdrahtet. Ohne Verdrahtung bewusst ein lauter Fehler (kein stiller Fallback, S3).


class DnsBypassViewRunner(Protocol):
    """Schmaler Vertrag des injizierten Lese-Runners (liefert die fertige Sicht)."""

    async def __call__(self) -> DnsBypassOverviewOut:
        """Baut die verdichtete netzweite Umgehungs-Sicht und liefert sie api-fertig."""
        ...


def provide_dns_bypass_view() -> DnsBypassViewRunner:
    raise NotImplementedError("DnsBypassViewRunner wird in app.py verdrahtet")


# Start-Runner: startet die Aufzeichnung ON-DEMAND fuer ein optionales Interface und gibt
# ``None`` bei Erfolg oder einen ehrlichen Fehlertext zurueck (der Helfer-Start kann
# legitim scheitern). Synchron -- der Recorder-Start ist ein lokaler Steuerbefehl.
type DnsBypassStartRunner = Callable[[str | None], str | None]


def provide_dns_bypass_start() -> DnsBypassStartRunner:
    raise NotImplementedError("DnsBypassStartRunner wird in app.py verdrahtet")


# Stop-Runner: stoppt die Aufzeichnung (idempotent/best-effort) -> None.
type DnsBypassStopRunner = Callable[[], None]


def provide_dns_bypass_stop() -> DnsBypassStopRunner:
    raise NotImplementedError("DnsBypassStopRunner wird in app.py verdrahtet")


class DnsBypassStatusRunner(Protocol):
    """Schmaler Vertrag des injizierten Status-Runners (billiger Poll, keine Verdichtung)."""

    def __call__(self) -> DnsBypassStatusOut:
        """Liefert den laufenden Zustand des Recorders (recording + collected_queries)."""
        ...


def provide_dns_bypass_status() -> DnsBypassStatusRunner:
    raise NotImplementedError("DnsBypassStatusRunner wird in app.py verdrahtet")


class StartBody(BaseModel):
    """POST /api/dns-bypass/start -- optionales Interface fuer die Aufzeichnung.

    ``interface`` ist ``None`` -> der Helfer waehlt sein Default-Interface (netzweiter
    Sniff mit ``promisc=True``, ADR 0042).
    """

    interface: str | None = None


# ── Routen ───────────────────────────────────────────────────────────────────


@router.get("")
async def get_dns_bypass(
    runner: Annotated[DnsBypassViewRunner, Depends(provide_dns_bypass_view)],
) -> DnsBypassOverviewOut:
    """Liefert die verdichtete netzweite DNS-Umgehungs-Sicht (Befunde + Zaehler + Status)."""
    return await runner()


@router.get("/status")
def get_dns_bypass_status(
    runner: Annotated[DnsBypassStatusRunner, Depends(provide_dns_bypass_status)],
) -> DnsBypassStatusOut:
    """Billiger Status-Poll: laeuft die Aufzeichnung? wie viele Anfragen sind gesammelt?"""
    return runner()


@router.post("/start")
async def start_dns_bypass(
    body: StartBody,
    runner: Annotated[DnsBypassStartRunner, Depends(provide_dns_bypass_start)],
) -> dict[str, object]:
    """Startet die Aufzeichnung ON-DEMAND (Muster pcap-Capture): Task erst bei Bedarf.

    Der Helfer-Start kann legitim scheitern (z. B. keine Rechte): ein Fehlertext wird
    ehrlich als ``{"ok": false, "error": <text>}`` mit HTTP 200 durchgereicht -- KEIN 500,
    kein stiller Fallback (S3). Erfolg (``None``) -> ``{"ok": true}``.

    BEWUSST ``async`` (Muster ``traffic.start_traffic_poll``): der Composition-Root-
    Runner ruft ``asyncio.create_task`` -- das braucht einen laufenden Event-Loop. Ein
    synchroner Endpunkt liefe im Starlette-Threadpool OHNE Loop (``RuntimeError: no
    running event loop``). Der ``async``-Endpunkt laeuft im Loop; der ``runner``-Callable
    selbst bleibt synchron und wird weiterhin OHNE ``await`` aufgerufen.
    """
    error = runner(body.interface)
    if error is not None:
        return {"ok": False, "error": error}
    return {"ok": True}


@router.post("/stop")
def stop_dns_bypass(
    runner: Annotated[DnsBypassStopRunner, Depends(provide_dns_bypass_stop)],
) -> dict[str, bool]:
    """Stoppt die Aufzeichnung (idempotent/best-effort). ``{"ok": true}``."""
    runner()
    return {"ok": True}

File: frontend/src/api/dnsBypass.js
// Netzweiter DNS-Umgehungs-Waechter: Mapper + REST-Aufrufe (CERNIS PRO 2.0, E5)
//
// Lese-/Steuerpfade der NETZWEITEN DNS-Umgehungs-Sicht (Reiter 2 des DNS-
// Waechters). Uebersetzt die Backend-Antworten (snake_case, aus
// backend/api/dns_bypass.py, Etappe 4b) in die camelCase-View-Struktur. Stil
// bewusst wie api/dnsWatch.js und api/outboundLog.js: kleine reine Mapper,
// null = ehrlich nicht vorhanden (KEIN erfundener Fallback), Fehler ueber
// ApiError. Pfade bleiben relativ; NIE einen Host hartkodieren.
//
// Wire-Form (verifiziert, aus backend/api/dns_bypass.py, NICHT aendern),
// Prefix /api/dns-bypass:
//   GET  ""        -> { findings:[ {src_ip,device_name,dst_ip,is_doh,
//                        doh_source_name,query_count,sample_qnames[]} ],
//                        expected_servers[], queries_total, bypass_total,
//                        expected_total, bypass_devices, recording }
//   GET  "/status" -> { recording, collected_queries, permission_error }
//   POST "/start" Body {interface?} -> {ok:true} | {ok:false, error:<text>}
//   POST "/stop"   -> {ok:true}
//
// Der Start-Helfer kann legitim scheitern (z. B. keine Rechte): das Backend
// reicht den Fehlertext als {ok:false, error} mit HTTP 200 durch (S3, kein
// stiller Fallback) -- startDnsBypass gibt genau dieses Objekt weiter, die View
// zeigt den Text ruhig an.
//
// KEINE Quittier-/Ack-Route: die 4b-Routen kennen keine Quittierung; hier wird
// daher bewusst KEINE gebaut (nur sichtbar machen, nicht markieren).

import { apiGet, apiPost } from "./client.js";

// Basis-Pfad der netzweiten Umgehungs-Ressource (relativ; Vite-Proxy leitet ans
// Backend).
const BASIS = "/api/dns-bypass";

// Ein Umgehungs-Befund-Wire-dict (snake_case) -> View-Objekt (camelCase). Je
// (srcIp, dstIp)-Gruppe eine Zeile. deviceName/dohSourceName bleiben ehrlich
// null, wenn sie fehlen (?? null) -- kein erfundener Name. isDoh ist eine
// heuristische Ziel-Bewertung (neutral, nicht gesichert). sampleQnames faellt
// auf [] zurueck, falls das Feld fehlt.
export function mappeBefund(f) {
  return {
    srcIp: f.src_ip,
    deviceName: f.device_name ?? null,
    // Ob die Quell-IP der eigene Host ist (Geraet mit source=self) -- treibt das
    // "eigener Host"-Badge in der Zeile. Fehlt das Feld -> false (ehrlich).
    isSelf: Boolean(f.is_self),
    dstIp: f.dst_ip,
    // Best-effort Ziel-Resolver-Name (Bestand > bekannte Resolver > PTR); fehlt er,
    // bleibt es bei der rohen IP (?? null -- kein erfundener Name).
    resolverName: f.resolver_name ?? null,
    isDoh: Boolean(f.is_doh),
    dohSourceName: f.doh_source_name ?? null,
    queryCount: f.query_count,
    sampleQnames: f.sample_qnames ?? [],
  };
}

// GET "" -> die verdichtete netzweite Umgehungs-Sicht. Uebersetzt findings in
// die View-Form und reicht Zaehler/expectedServers/recording durch. Leere/
// fehlende Felder fallen ehrlich auf 0/[]/false zurueck (kein erfundener Wert).
// recording sagt ehrlich, ob gerade mitgelesen wird -- ein Leerbefund bei
// recording=false bedeutet: es wird NICHT mitgelesen, nicht "alles sauber".
export async function fetchDnsBypass() {
  const backend = await apiGet(BASIS);
  return {
    findings: (backend.findings ?? []).map(mappeBefund),
    expectedServers: backend.expected_servers ?? [],
    queriesTotal: backend.queries_total ?? 0,
    bypassTotal: backend.bypass_total ?? 0,
    expectedTotal: backend.expected_total ?? 0,
    bypassDevices: backend.bypass_devices ?? 0,
    recording: Boolean(backend.recording),
  };
}

// GET "/status" -> billiger Status-Poll (laeuft die Aufzeichnung? wie viele
// Anfragen sind gesammelt? traegt die Plattform die Erfassung?), ohne die teure
// Verdichtung. collectedQueries faellt auf 0 zurueck, falls das Feld fehlt.
//
// permissionError traegt denselben stabilen Marker wie der SNI-Status: liegt er
// an, ist die Erfassung auf dieser Plattform gar nicht moeglich und die Ansicht
// graut sie aus, statt den Anwender starten zu lassen. Fehlt das Feld oder ist
// es leer -> null (ehrlich "kein Hindernis bekannt", nicht "" als Marker).
export async function fetchDnsBypassStatus() {
  const backend = await apiGet(`${BASIS}/status`);
  return {
    recording: Boolean(backend.recording),
    collectedQueries: backend.collected_queries ?? 0,
    permissionError: backend.permission_error || null,
  };
}

// POST "/start" -> startet die Aufzeichnung ON-DEMAND. interface ist optional
// (null -> das Backend waehlt sein Default-Interface). Das Backend reicht einen
// legitimen Start-Fehler als {ok:false, error:<text>} mit HTTP 200 durch (S3,
// kein 500) -- dieses Objekt wird UNVERAENDERT weitergegeben, damit die View
// zwischen Erfolg ({ok:true}) und ehrlichem Fehlertext unterscheiden kann.
export async function startDnsBypass(iface = null) {
  const body = {};
  if (iface !== null && iface !== undefined) {
    body.interface = iface;
  }
  return apiPost(`${BASIS}/start`, body);
}

// POST "/stop" -> stoppt die Aufzeichnung (idempotent/best-effort). {ok:true}.
export async function stopDnsBypass() {
  return apiPost(`${BASIS}/stop`, {});
}

export default {
  mappeBefund,
  fetchDnsBypass,
  fetchDnsBypassStatus,
  startDnsBypass,
  stopDnsBypass,
};
